File: csharp/factory.cs
using System;

abstract class Animal {
    public abstract string Speak();
}

class Dog : Animal {
    public override string Speak() {
        return "Woof!";
    }
}

class Cat : Animal {
    public override string Speak() {
        return "Meow!";
    }
}

class AnimalFactory {
    public Animal CreateAnimal(string animalType) {
        if (animalType == "dog") {
            return new Dog();
        } else if (animalType == "cat") {
            return new Cat();
        }
        return null;
    }
}

File: kotlin/factory.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak(): String {
        return "Woof!"
    }
}

class Cat : Animal() {
    override fun speak(): String {
        return "Meow!"
    }
}

class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}